proposals/intraprocess_communication.md

```diff
@@ -0,0 +1,582 @@
+a ring buffer of the size specified by the depth from the QoS.
+
+### Creating a subscription
+
+1. User calls `Node::create_subscription<MessageT>(...)`.
+2. This boils down to `NodeTopics::create_subscription(...)`, where a `Subscription` is created through the factory.
+3. Here, if intra-process communication is enabled, intra-process related variables are initialized through the
+   `Subscription::SetupIntraProcess(...)` method. The most relevant ones being the ring buffer and the waitable object.
+4. Then the `IntraProcessManager` is notified about the existence of the new `Subscription` through the method
+   `IntraProcessManager::add_subscription(SubscriptionBase::SharedPtr subscription, SubscriptionOptions options)`.
+5. `IntraProcessManager::add_subscription(...)` stores the `Subscription` information in an internal structure
+   of type `SubscriptionInfo`. The structure contains information about the `Subscription`, such as its QoS, its topic name
+   and the type of its callback, and a weak pointer for the `Subscription` object.
+   An `uint64_t sub_id` unique within the `rclcpp::context` is assigned to the `Subscription`.
+   The `IntraProcessManager` contains a `std::map<uint64_t, SubscriptionInfo>` object where
+   it is possible to retrieve the `SubscriptionInfo` of a specific `Subscription` given its id.
+   There is also an additional structure `std::map<uint64_t, std::pair<std::set<uint64_t>, std::set<uint64_t>>>`.
+   The key of the map is the unique id of a `Publisher`. The value of the map is a pair of sets of ids. These
+   sets contain the ids of the `Subscription`s that can communicate with the `Publisher`. We have two
+   different sets because we want to differentiate the `Subscription`s depending on whether they request
+   ownership of the received messages or not.
+6. The `SubscriptionIntraProcessWaitable` object is added to the list of Waitable interfaces of the node through
+   `node_interfaces::NodeWaitablesInterface::add_waitable(...)`.
+
+The following steps will be executed if the `Subscription` QoS is set to `Transient Local`.
+In this case the `IntraProcessManager` has to check if the recently created `Subscription` is a late-joiner, and in that case,
+retrieve messages from the `Transient Local` `Publisher`s.
+
+7. Call `IntraProcessManager::find_matching_publishers(SubscriptionInfo sub_info)` that returns a list
+   of stored `PublisherInfo` that have a QoS compatible for sending messages to this new `Subscription`.
+8. Check if any of these `Publisher` have a transient local QoS. If this is the case, they will have a ring
+   buffer.
+9. Copy messages from all the ring buffers found into the ring buffer of the new `Subscription`. **TODO:** are
+   there any constraints on the order in which old messages have to be retrieved? (i.e. 1 publisher at the
+   time; all the firsts of each publisher, then all the seconds ...).
```

Comment:
Oh yes, I updated saying that if a compatible publisher is found it must be a transient local publisher.
Implementing "best available durability" for the intra-process communication is simple, but I guess that is not desirable since it will introduce a relevant difference between when intra-process is enabled or not.